bip-0352.mediawiki

```diff
@@ -0,0 +1,457 @@
+* Exclude inputs with witness version > 1 (see ''[[#scanning-transactions|Scanning transactions]]'')
+* For each taproot output spent the sending wallet MUST have access to the private key corresponding to the taproot output key, unless ''H'' is used as the internal public key
+
+==== Creating outputs ====
+
+After the inputs have been selected, the sender can create one or more outputs for one or more silent payment addresses in the following manner:
+
+* Collect the private keys for each input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* For each private key ''a<sub>i</sub>'' corresponding to a [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot output, check that the private key produces a point with an even y-value and negate the private key if not<ref name="why_negate_taproot_private_keys">'''Why do taproot private keys need to be checked?''' Recall from [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] that each x-only public key has two corresponding private keys, ''d'' and ''n - d''. To maintain parity between sender and receiver, it is necessary to use the private key corresponding to the even y-value when performing the ECDH step since the receiver will assume the even y-value when summing the taproot x-only public keys.</ref>
+* Let ''a = a<sub>0</sub> + a<sub>1</sub> + … a<sub>n</sub>'', where each ''a<sub>i</sub>'' has been negated if necessary
+* Generate the ''outpoints_hash'', using the method described above
+* Group receiver silent payment addresses by ''B<sub>scan</sub>'' (e.g. each group consists of one ''B<sub>scan</sub>'' and one or more ''B<sub>m</sub>'')
+* For each group:
+** Let ''ecdh_shared_secret = outpoints_hash·a·B<sub>scan</sub>''
+** Let ''n = 0''
+** For each ''B<sub>m</sub>'' in the group:
+*** Let ''t<sub>n</sub> = sha256(ser<sub>P</sub>(ecdh_shared_secret) || ser<sub>32</sub>(n))''
+*** Let ''P<sub>mn</sub> = B<sub>m</sub> + t<sub>n</sub>·G''
+*** Encode ''P<sub>mn</sub>'' as a [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot output
+*** Optionally, repeat with n++ to create additional outputs for the current ''B<sub>m</sub>''
+*** If no additional outputs are required, continue to the next ''B<sub>m</sub>'' with ''n++''<ref name="why_not_the_same_tn">''' Why not re-use ''t<sub>n</sub>''? when paying different labels to the same receiver?''' If paying the same entity but to two separate labeled addresses in the same transaction without incrementing ''n'', the two outputs would be ''B<sub>spend</sub> + t<sub>n</sub>·G + i·G'' and ''B<sub>spend</sub> + t<sub>n</sub>·G + j·G''. The attacker could subtract the two values and observe that the distance between i and j is small. This would allow them to deduce that this transaction is a silent payment transaction and that a single entity received two outputs, but won't tell them who the entity is.</ref>
```

Comment:
Thinking about it if labels are not small incremental int as I suggested above we may reuse _tn_. Not sure if it makes a lot of difference in practice though, I don't think it will often happen that a sender wants to pay _that_ many labels of the same receiver in one transaction so that not reusing _tn_ would induce a noticable burden.

Reply:
That's a good point, however, the sender has no idea what kind of label scheme the receiver has used, so it's possible their transaction could be doxxed as a silent payment transaction if the receiver uses an incremental integer scheme. Also worth pointing out that hashing the integer counter for each new output is a relatively cheap operation, and doing so gives us a guarantee that the transaction won't leak information, regardless of the labeling scheme used by the receiver

Reply:
Agree, probably best as it is now.